langchain/retrievers/perfect_retriever.py

```diff
@@ -0,0 +1,220 @@
+import json
+from abc import abstractmethod
+from typing import Any, Dict, List
+
+from langchain import LLMChain, PromptTemplate
+from langchain.agents import AgentExecutor
+from langchain.agents.agent_toolkits import VectorStoreInfo
+from langchain.llms import BaseLLM
+from langchain.schema import BaseOutputParser, BaseRetriever, Document
+from langchain.vectorstores.base import VectorStoreRetriever
+
+pinecone_format_instructions = """RESPONSE FORMAT
+----------------------------
+
+When responding to me please, please output a response in this format. Use a markdown code snippet formatted in the following schema:
+
+```json
+{{{{
```

Comment:
double brackets is how you escape brackets for python str formatting

{ = formatted first time you call .format, {{ = formatted second time, {{{{ = formatted third time. this string will be added to overall prompt template and then format will be called twice (once to add the description of the doc store and metadata fields, once to enter the actual user query), so that by the time this is passed to llm it'll be reduced to a single { (so it'll look like json)

so 
```python
x = "{{{{foo_bar}}}}
x.format() == "{{foo_bar}}"
x.format().format() == "{foo_bar}"
x.format().format().format(foo_bar="whatever") == "whatever"
```